### ISSUE TITLE
Stop using CDN to provide website blocked by firewall

### ISSUE THREAD

**Issue by kinosang** [CacheBrowser/cachebrowser#2]
It's not a good idea to "go pass" firewall. Due to the same reason, GFW has blocked or jammed many CDN providers such as AWS CloudFront/S3, Google PageSpeed, Cloudflare, incapsula, etc.

**Comment by legendtang** [comment]
```
[HTTPS] www.google.com.hk:443  <PROXYING>
Traceback (most recent call last):
  File "build\bdist.win-amd64\egg\cachebrowser\common.py", line 52, in inner
    return func(*args, **kwargs)
  File "build\bdist.win-amd64\egg\cachebrowser\proxy.py", line 45, in on_local_data
    self._schema = schema(self, self._buffer)
  File "build\bdist.win-amd64\egg\cachebrowser\proxy.py", line 173, in __init__
    self._start_upstream()
  File "build\bdist.win-amd64\egg\cachebrowser\proxy.py", line 211, in _start_upstream
    return self._connect_upstream(host, port)
  File "build\bdist.win-amd64\egg\cachebrowser\proxy.py", line 225, in _connect_upstream
    sock.connect((ip, port))
  File "D:\Python27\lib\site-packages\gevent-1.1rc1-py2.7-win-amd64.egg\gevent\_socket2.py", line 220, in connect
    raise error(err, strerror(err))
error: [Errno 10060] [Error 10060] 由于连接方在一段时间后没有正确答复或连接的主机没有反应，连接尝试失败。

[HTTPS] www.google.com.hk:443  <PROXYING>
[HTTPS] www.google.com.hk:443  <PROXYING>
```

Agreed as @kinosang said. Your project is not like the same thing stated in your thesis that works like a charm. You must consider more about realistic Chinese "internet" environment, such as which CDN providers are to use.

However in my prospect, those CDNs that are not jammed don't cache the content blocked by China at present. :/

**Comment by kinosang** [comment]
@legendtang Not only China, other countries with Internet censorship may also block or jam CDNs. CDN is designed for content delivering, not for anti-censorship.

**Comment by xgdgsc** [comment]
I feel this would cause more trouble for people who regularly use sites that are not blocked right now, and don' t care accessing blocked sites.  This might end up with failing to appeal the minority while having a high risk of sacrificing the benefits of the majority.

**Comment by ghost** [comment]
This issue is absurd. You are telling the creator to stop the exact thing the program is designed to do. Unless you have a better idea how to work around a firewall you should drop it.

**Comment by xgdgsc** [comment]
This issue isn' t absurd. Thinking oneself doing something helpful while might actually harmful  is absurd.

**Comment by ghost** [comment]
@xgdgsc encouraging people to stop fighting for freedom because of the threat of further censorship is the definition of fascism.

**Comment by legendtang** [comment]
@svnpenn No, we do have better ones to work around it even if this is a passable method. Do more investigates before complaining about our resistance.

This is not about losing freedom or faith. I'm just questioning the feasibility of this idea when facing the reality.

According to my research, the program fails to determine whether a blocked website is through bootstrapped CDNs. The hard-code arrays of original hosts & CDNs won't change for now.

**Comment by Zola** [comment]
CDN provider have rights to select customer . we can use it before CDN provider say no.

**Comment by ghost** [comment]
@legendtang you do know how Git works right? It has these things called "commits".

**Comment by z3t1** [comment]
@xgdgsc The majority don't care about censorship at all. They only use domestic web services, not even aware of the wall. 

Those who complain about the absurd issue are people like you who want to cross the wall but aren't willing to fight.

**Comment by legendtang** [comment]
@svnpenn I clearly stated "for now". And I do think I should open another issue to discuss those I had pointed out.

This issue is discussing about feasibility of this idea in unblocking method mentioned in the thesis before those commits are coming. Please be respectful.

**Comment by ghost** [comment]
@legendtang questioning the existence of a repository, rather than problems with the code. Might I present you with a mirror?

**Comment by z3t1** [comment]
@legendtan Usability certainly is another topic. You should open your own issue.

**Comment by Artoria2e5** [comment]
I consider using CDNs a new and interesting method to go through not-that-carefully designed censored environments. This tool provides an easy way to transfer the damage caused by censorship attempts to something bigger -- blocking CDNs is always a big and risky thing to do, and not everyone wants to take such risks.

Censorship is not limited to big govs like the PRC one. Schools and other workplaces often give censored Internet, and they can't afford the aftermath of blocking CDNs, neither do _sane_ govs. Look around and recognize the existence of technically weaker adversaries like meraki.

(Ah, yes, I use TOR at school……)

---

Hmm, what license does the tool use? Just add some **ABSOLUTELY NO WARRANTY** and give this issue a wontfix.

---

> You go everywhere to follow the big news, but the questions you ask are too simple — sometimes naïve. Understand, or not?
> 
> [...] need to raise your level of knowledge. Got it, or not? I'm anxious for you all, it's true. You really, I.
> 
> 你们这样子啊是不行的，我今天算得罪了你们一下！ -- [Jiang](https://zh.wikiquote.org/wiki/%E6%B1%9F%E6%B3%BD%E6%B0%91) [Zemin](https://en.wikiquote.org/wiki/Jiang_Zemin)

**Comment by superfashi** [comment]
I think the government won't risk to block CDNs as  there're also domestic websites using those.

**Comment by samos123** [comment]
This idea could be built upon by letting everybody become part of some kind of distributed CDN. Living in China I confirm that CDNs will and some are already blocked, you need to make the assumption in CacheBrowser that any CDN may get blocked and handle this case accordingly, if not your tool won't succeed.

Anyway even if CDNs will get blocked this is a good thing as Chinanet will get more isolated and hopefully more people will start to notice the government isn't providing GFW for the benefits of the people, but only use GFW to keep the government in power and to crack down any kind of resistance. Having the people responsible for GFW continually working their ass off to keep it up to date with latest tech is also a plus. Saying this will hurt the people is bullshit. The people who want to get out the GFW will get out with or without blocked CDNs.

**Comment by kinosang** [comment]
@svnpenn If all the global CDNs were blocked, Govs can keep all websites outside their contries under censorship or just jam all website using CDNs.

This issue is not for stoping people fighting for freedom, I just want to save possibility for people to get information from Internet.

Some websites have tried to use CDNs to go pass GFW, but Chinese gov just blocked more CDN providers.

So all of us, developers from China, know that use CDNs to provide information gov don't like is such a bad idea. CDNs is not designed for it.

**Comment by ghost** [comment]
@kinosang your entire argument is flawed. You assume that if we "play nice" and don't develop tools such as this, the government will leave CDNs alone. False. You **hope** that they will leave them alone. Even if this tool is not developed, they may choose to block all CDN for other reasons.

This tool should exist. We win by fighting, not by sitting idly by.

**Comment by ghost** [comment]
@kinosang From my point of view, you are just finding excuse for your own cowardly behaviours.

GFW is destroying the future of your country. When things like this happen, you should not compromise, you fight if you want to win. Or you might as well just leave it and let others do. Don't come make a fuss about people doing real sh*t which you can not possibly even think about. Shame on you.

> 苟利國家生死以，豈因禍福避趨之。

If the terrorists blew up a stadium and killed thousands. You don't blame the people gathering in the stadium that made it a target. You blame the terrorists. GFW is just a sort of terrorism IMO, which virtually does far more damage than those occasional random bomb attacks by making a whole generation of people blind and ignorant.

**Comment by neruthes** [comment]
It's quite reasonable to assume that Internet censorship in China is on an aggressive basis, not a passive one. China government will soon (in decades or scores) shut down access to Internet and build its own Chinanet, then definitely turn its people into slavery.
It deserves a lot to fight against the nation, even by sacrificing interest of some of its people. Every attempt and effort to destroy China shall be justice. Even if I will retain China citizenship and residency by the day. I'm willing to be sacrificed for the greater good, while I won't oblige everyone to think the same. But, be aware that, you cannot protect yourself and your siblings indefinitely under the sovereignty by compromising and enduring. If we don't destroy it, it will slave all people into its absolute tyranny.

**Comment by kinosang** [comment]
@svnpenn @Elethom Maybe you're right.

**Comment by Yaiba** [comment]
Maybe this project is not suitable for China? So what's the suitable case?
